osbuild/meta.py

```diff
@@ -585,3 +623,55 @@ def get_schema(self, klass, name=None, version="1") -> Schema:
         self._schemata[(klass, name, version)] = schema
 
         return schema
+
+    def list_runners(self, distro: Optional[str] = None) -> List[RunnerInfo]:
+        """List all available runner modules
+
+        The list is sorted by distribution and version (ascending).
+        If `distro` is specified, only runners matching that distro
+        will be returned.
+        """
+        path = os.path.join(self.path, "runners")
+        names = filter(lambda f: os.path.isfile(f"{path}/{f}"),
+                       os.listdir(path))
+        paths = map(lambda n: os.path.join(path, n), names)
+        mapped = map(RunnerInfo.from_name, paths)
+        runners = sorted(mapped, key=lambda r: (r.distro, r.version))
+
+        if distro:
+            runners = [r for r in runners if r.distro == distro]
+
+        return runners
+
+    def detect_runner(self, name) -> RunnerInfo:
+        """Detect the runner for the given name
+
+        Name here refers to the combination of distribution with an
+        optional version suffix, e.g. `org.osbuild.fedora30`.
+        This functions will then return the best existing runner,
+        i.e. a candidate with the highest version number that
+        fullfils the following criteria:
+          1) distribution of the candidate matches exactly
+          2) version of the candidate is smaller or equal
+        If no such candidate exists, a `ValueError` will be thrown.
+        """
+        name, version = RunnerInfo.parse_name(name)
+        candidate = None
+
+        # Get all candidates for the specified distro (1)
+        candidates = self.list_runners(name)
+
+        while candidates:
+            # NB: we consume the list in reverse order
+            candidate = candidates.pop()
+
+            if candidate.version == version:
+                return candidate
+            if candidate.version < version:
+                break
```

Comment:
```python
if candidate.version <= version
    return candidate
```

would this be the same logic?  I think it would be clearer than breaking on `<` and returning after.

Reply:
Probably leftover from some other thinking that included that fallback. Will simplify.